### PR TITLE
fix: use skill references in SKILL.md for skills-only delivery

### LINKED ISSUE #879: [Bug] Skills-only delivery emits `/opsx:*` command references
# OpenSpec Bug Report: Skills-only delivery emits `/opsx:*` command references

**Date**: 2026-03-26
**OpenSpec version**: 1.2.0
**Context**: Discovered while auditing `.claude/` files in a project configured with `delivery: 'skills'`. All generated openspec-* skills contained `/opsx:*` cross-references pointing to commands that were never generated.

---

## Bug title

**Skills-only delivery generates skill files with `/opsx:*` command references instead of `/openspec-*` skill references**

## Description

When OpenSpec is configured with `delivery: 'skills'` (skills only, no commands), the generated SKILL.md files contain hardcoded `/opsx:*` cross-references (e.g., `/opsx:apply`, `/opsx:archive`, `/opsx:verify`). These reference commands that don't exist in a skills-only setup, creating broken workflows.

## Expected behavior

When `delivery: 'skills'`, cross-references in generated skills should use the skill namespace: `/openspec-apply-change`, `/openspec-archive-change`, `/openspec-verify-change`, etc.

## Actual behavior

All cross-references use the command namespace (`/opsx:apply`, `/opsx:archive`, etc.) regardless of the delivery setting.

## Affected skills

Every skill with cross-references is affected:

| Skill | Broken references found |
|-------|------------------------|
| `openspec-propose` | `/opsx:apply` |
| `openspec-ff-change` | `/opsx:apply` |
| `openspec-apply-change` | `/opsx:apply`, `/opsx:archive` |
| `openspec-explore` | `/opsx:explore` |

## Root cause

1. **Skill templates hardcode `/opsx:*`**: The workflow template functions in `src/core/templates/workflows/*.ts` (e.g., `getOpsxProposeSkillTemplate`, `getApplyChangeSkillTemplate`) embed `/opsx:*` command references directly in the `instructions` strings.

2. **No delivery-aware transformer for skills**: `generateSkillContent()` in `src/core/shared/skill-generation.ts` accepts a `transformInstructions` callback, but:
   - The only transformer is `transformToHyphenCommands` (for OpenCode: `/opsx:` → `/opsx-`)
   - No transformer exists for the skills-only case (`/opsx:*` → `/openspec-*`)
   - In `src/core/init.ts` and `src/core/update.ts`, when `delivery === 'skills'`, no transformer is passed

## Suggested fix

The mapping isn't a simple regex — command short names differ from skill directory names:

| Command ref | Skill ref |
|-------------|-----------|
| `/opsx:apply` | `/openspec-apply-change` |
| `/opsx:archive` | `/openspec-archive-change` |
| `/opsx:verify` | `/openspec-verify-change` |
| `/opsx:continue` | `/openspec-continue-change` |
| `/opsx:propose` | `/openspec-propose` |
| `/opsx:explore` | `/openspec-explore` |
| `/opsx:ff` | `/openspec-ff-change` |
| `/opsx:new` | `/openspec-new-change` |

Options:
1. Add a `transformToSkillReferences` function in `src/utils/command-references.ts` with the explicit mapping table above. Apply it in `init.ts` and `update.ts` when `delivery === 'skills'`.
2. Or refactor templates to accept a reference-format parameter, emitting correct cross-references at construction time.

## Reproduction

1. Run `openspec init` and choose `delivery: 'skills'`
2. Inspect any generated skill, e.g., `.claude/skills/openspec-propose/SKILL.md`
3. Observe `/opsx:apply` in the output text — a command that was never generated

### LINKED ISSUE #881: [Fix] Add transformToSkillReferences for skills-only delivery mode
## Context

Reported in #879 by @m9dfukc. When `delivery: 'skills'` is configured, generated SKILL.md files contain hardcoded `/opsx:*` command references that point to commands that don't exist in skills-only mode.

## Root Cause

- Workflow templates in `src/core/templates/workflows/*.ts` hardcode `/opsx:*` references (46 total across all templates)
- `generateSkillContent()` supports a `transformInstructions` callback, but only `transformToHyphenCommands` (for OpenCode) exists
- `init.ts` and `update.ts` don't pass any transformer when `delivery === 'skills'`

## Proposed Fix

1. Add `transformToSkillReferences()` to `src/utils/command-references.ts` using the explicit mapping from `WORKFLOW_TO_SKILL_DIR` in init.ts:
   - `/opsx:apply` → `/openspec-apply-change`
   - `/opsx:archive` → `/openspec-archive-change`
   - `/opsx:verify` → `/openspec-verify-change`
   - `/opsx:continue` → `/openspec-continue-change`
   - `/opsx:propose` → `/openspec-propose`
   - `/opsx:explore` → `/openspec-explore`
   - `/opsx:ff` → `/openspec-ff-change`
   - `/opsx:new` → `/openspec-new-change`
   - `/opsx:sync` → `/openspec-sync-specs`
   - `/opsx:bulk-archive` → `/openspec-bulk-archive-change`
   - `/opsx:onboard` → `/openspec-onboard`

2. Wire the transformer in `src/core/init.ts` and `src/core/update.ts`:
```typescript
const transformer = tool.value === 'opencode'
  ? transformToHyphenCommands
  : delivery === 'skills'
    ? transformToSkillReferences
    : undefined;
```

3. Add tests for the new transformer

## Files to Modify

- `src/utils/command-references.ts` — add `transformToSkillReferences()`
- `src/core/init.ts` — wire transformer (~line 541)
- `src/core/update.ts` — wire transformer (~lines 198, 669)
- Tests

Closes #879

### PR DESCRIPTION
**Pull request by mc856**
Fixes #881, closes #879.

When `delivery: 'skills'` is configured, generated SKILL.md files contained hardcoded `/opsx:*` command references pointing to commands that were never generated. This implements the fix outlined in #881: adds `transformToSkillReferences()` with the explicit 11-entry command-to-skill mapping and wires it at every `generateSkillContent` call site — `init.ts`, both sites in `update.ts` (including `upgradeLegacyTools`), and both sites in `workspace/skills.ts`.

## Deviations from the issue snapshot

- **Kept the existing `pi` branch** in the transformer condition. The issue snippet only checks `opencode`, but main already routes `pi` through `transformToHyphenCommands`; copying the snippet verbatim would regress pi.
- **`workspace/skills.ts` wired in addition to the files listed in the issue.** It generates SKILL.md from the same `globalConfig.delivery` and is itself a skills-only surface, so the bug persisted there. Regression test included (verified to fail against the unwired code).
- Note: `WORKFLOW_TO_SKILL_DIR` exists in both `init.ts` (local copy) and `profile-sync-drift.ts` (exported); the two are identical and `COMMAND_TO_SKILL_REFERENCE` matches both.

## Open question for maintainers

In skills-only mode, opencode/pi still take the `transformToHyphenCommands` branch, so their SKILL.md files contain `/opsx-*` references to commands that were also never generated — same class of bug. I didn't expand scope because it's unclear whether `/openspec-<skill>` slash references resolve in those tools. Happy to file a follow-up issue. The legacy-upgrade path also lacks a skills-only content assertion; same offer.

## Testing

- `npx eslint src/` clean
- `transformToSkillReferences` unit tests: per-mapping assertions for all 11 commands, multi-reference, backtick, unknown-command, and bulk-archive/archive boundary cases
- Integration: init + update skills-only cases assert generated SKILL.md contains no `/opsx:` and does contain `/openspec-` (explore template verified to contain `/opsx:` references, so the assertion is discriminative); new workspace test mutation-verified
- Full vitest suite passing

Generated with Claude (Cowork) using claude-fable-5.

## Summary by CodeRabbit

* **Bug Fixes**
  * Skill setup and update flows now generate the correct command/skill references for skills-only delivery.
  * Generated `SKILL.md` files no longer include unsupported `/opsx:` markers and instead link to the expected `/openspec-` skill references.
  * Reference rewriting is now chosen consistently by delivery mode, improving output across supported tools.
* **Tests**
  * Expanded coverage to assert the correct reference markers in init and update `SKILL.md` outputs, plus command reference transformation behavior.

### CODE REVIEW

**Comment by mc856** [comment]
Pushed 4ef4c48 addressing the CodeRabbit review: extracted the duplicated transformer selection into `getTransformerForTool()` in `command-references.ts` (with JSDoc and unit tests locking the selection matrix), replacing all five call sites.

On consolidating the `WORKFLOW_TO_SKILL_DIR` copies: agreed in principle, but deriving the map in `src/utils` from `src/core/profile-sync-drift.ts` would invert the utils ← core layering, and moving the canonical map into utils touches `profile-sync-drift.ts` typing and `init.ts` beyond this bugfix's scope. Happy to do it as a follow-up PR if maintainers agree.

**Review by alfred-openspec — CHANGES_REQUESTED** [review]
Thanks for tightening this up. The Claude/init/update/workspace path is in much better shape, but I found one remaining skills-only edge case before this can merge.

`getTransformerForTool()` currently returns `transformToHyphenCommands` for OpenCode and Pi before it checks `delivery === 'skills'`. That means `delivery: 'skills'` still generates OpenCode/Pi SKILL.md files that point at `/opsx-*` commands even though command files were removed or never generated. For skills-only delivery, the missing-command fix needs to win for every tool, including OpenCode and Pi. Please make `delivery === 'skills'` return skill references first, then keep the hyphen transform for OpenCode/Pi when commands are actually generated, and add focused coverage for those two tools.

Checked: focused command-reference/init/update/workspace tests pass locally.

**Comment by mc856** [comment]
Done in 8c66688 — `delivery === 'skills'` now wins for every tool: `getTransformerForTool` checks skills-only delivery first and only falls back to the hyphen transform for opencode/pi when commands are actually generated. Added unit coverage for the opencode/pi selection matrix and an opencode skills-only init integration test asserting the generated SKILL.md contains no `/opsx:` or `/opsx-` references. This also closes out the open question from the PR description — thanks for the call.

**Review by alfred-openspec — DISMISSED** [review]
Approving latest head. I verified the skills-only transformer now wins for every tool, including OpenCode/Pi, and init/update/workspace skill generation have focused regression coverage. Targeted install, tsc, lint on src, and the relevant Vitest suites all pass.

**Comment by mc856** [comment]
@alfred-openspec I rebased this onto the latest upstream main and resolved the merge conflict; GitHub now reports it as mergeable and CodeRabbit is green again. The previous approval was marked stale after the force-push, so could you please re-confirm when you have a chance?

**Review by alfred-openspec — DISMISSED** [review]
Re-approved after the rebase. I rechecked the diff: skills-only delivery now emits skill references for all tools, OpenCode/Pi precedence is covered, and the affected init/update paths plus transformer helper have focused regression tests. Verified locally with install, tsc, eslint on src, and targeted Vitest suites.

**Comment by mc856** [comment]
I rebased this onto the latest upstream main and resolved the merge conflicts from the v1.6.0 changes — the transformer selection now also covers the newly added oh-my-pi tool, with matching test coverage. GitHub reports the PR as mergeable again. I also added docstrings to the touched entry points to clear the docstring coverage warning. @alfred-openspec the rebase marked your approval as stale — could you re-confirm when you get a chance?

File: test/core/init.test.ts [hunk 1]
```diff
@@ -742,6 +742,31 @@ describe('InitCommand - profile and detection features', () => {
     // Commands should NOT exist
     const cmdFile = path.join(testDir, '.claude', 'commands', 'opsx', 'explore.md');
     expect(await fileExists(cmdFile)).toBe(false);
+
+    // Skill content should reference skills, not commands that were never generated
+    const skillContent = await fs.readFile(skillFile, 'utf-8');
+    expect(skillContent).not.toContain('/opsx:');
+    expect(skillContent).toContain('/openspec-');
+  });
```
[the remark this answers is not in the document]
Comment (mc856) [reply test/core/init.test.ts:761]: Added — the Claude assertion now also rejects the stale hyphenated form — 9125398

File: test/core/update.test.ts [hunk 2]
```diff
@@ -1486,6 +1486,14 @@ More user content after markers.
...
         path.join(commandsDir, 'explore.md')
       )).toBe(false);
+
+      // Skill content should reference skills, not commands that were never generated
+      const skillContent = await fs.readFile(
+        path.join(skillsDir, 'openspec-explore', 'SKILL.md'),
+        'utf-8'
+      );
+      expect(skillContent).not.toContain('/opsx:');
+      expect(skillContent).toContain('/openspec-');
```
[the remark this answers is not in the document]
Comment (mc856) [reply test/core/update.test.ts:1497]: Added — the update assertion now also rejects the stale hyphenated form — 9125398

**Comment by clay-good** [comment]
Nice fix, and the opencode/pi precedence is now handled correctly. One gap before this is complete:

The new `COMMAND_TO_SKILL_REFERENCE` map is missing the `update` command. It has 11 entries, but the canonical `WORKFLOW_TO_SKILL_DIR` (which the map's own comment says to stay in sync with) has 12 — it also includes `update`.

Why it matters: `update` is a **core workflow**, so `openspec-update-change/SKILL.md` is generated by default, and that template references `/opsx:update` (src/core/templates/workflows/update-change.ts:104). In skills-only mode the transformer leaves unknown commands untouched, so that skill still ships a broken `/opsx:update` reference — the exact bug this PR fixes, just for one more command.

The current tests miss it because they only assert against the `explore` skill, which has no `/opsx:update`.

Suggested fix:
- Add `'update': '/openspec-update-change',` to `COMMAND_TO_SKILL_REFERENCE`.
- Add an assertion against the `openspec-update-change` skill (or any skill containing an update reference) so `not.toContain('/opsx:')` would catch a future omission.

Everything else looks good to go.

**Comment by mc856** [comment]
Good catch — added the `update` entry in bb81a67, so the map now mirrors all 12 entries of `WORKFLOW_TO_SKILL_DIR`. One note: the skill variant of update-change drops the `/opsx:update` self-reference (only the command template at update-change.ts:104 carries it), so no generated skill currently ships the raw reference — the gap was latent. Covered it anyway: a unit test asserting the `/opsx:update` mapping, plus `not.toContain('/opsx:')` assertions on the generated `openspec-update-change` skill in both the init and update suites, so a future template that does reference update can't ship untransformed.

**Review by alfred-openspec — APPROVED** [review]
Verified at bb81a67: the skill-reference map now matches all 12 canonical workflows, including update, and skills-only delivery wins over hyphen command references for every tool. The 172 focused transformer, init, and update tests pass locally, and the full cross-platform CI suite is green.
